tests/integration-tests/tests/storage/kms_key_factory.py

```diff
@@ -1,136 +1,46 @@
-import json
 import logging
 import random
 import string
-import time
 
 import boto3
 import pkg_resources
 from jinja2 import Environment, FileSystemLoader
+from utils import get_arn_partition
 
 
 class KMSKeyFactory:
     """Manage creation for kms key."""
 
-    def __init__(self):
-        self.iam_client = None
+    def __init__(self, policies_factory, role_factory):
+        self.policies_factory = policies_factory
+        self.role_factory = role_factory
         self.kms_client = None
-        self.kms_key_id = None
-        self.account_id = None
-        self.region = None
-        self.partition = None
-        self.iam_role = None
-        self.iam_policy_arn_batch = None
-        self.iam_policy_arn_traditional = None
+        self.key_and_role_by_region = {}
 
     def create_kms_key(self, region):
         """
         Create a kms key with given region.
         :param region: Different region need to create different keys
         """
-        self.region = region
-        self.account_id = (
+        if region not in self.key_and_role_by_region:
+            policies = self.policies_factory(region)
+            user_iam_role = self.role_factory(
+                region, "ec2", [policies["awsbatch_instance_policy"], policies["traditional_instance_policy"]]
+            )
+            self.kms_key_id = self._create_kms_key(region, user_iam_role)
+            self.key_and_role_by_region[region] = (self.kms_key_id, user_iam_role)
```

Comment:
minor: it will be more clear for me to understand it is the combination of key id and role name if we name it  `self.key_id_and_role_by_region`, since we have lots of role arn, role name, key name, key id here